### PR TITLE
pipeline-manager: implement `/metrics` endpoint

### PR DESCRIPTION
**Pull request by abhizer**
Introduces a `/metrics` endpoint that returns pipeline metrics for all running pipelines owned by this tenant.

### CODE REVIEW

File: crates/pipeline-manager/src/api/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    tenant_id: ReqData<TenantId>,
+) -> Result<HttpResponse, ManagerError> {
+    let pipelines = state.db.lock().await.list_pipelines(*tenant_id).await?;
+
+    const NEWLINE: u8 = b'\n';
+
+    let mut result = Vec::new();
+
+    for pipeline in pipelines {
+        if pipeline.deployment_status == PipelineStatus::Running {
```
Comment (ryzhyk) [inline crates/pipeline-manager/src/api/metrics.rs:42]:
```suggestion
        if pipeline.deployment_status == PipelineStatus::Running || pipeline.deployment_status == PipelineStatus::Paused {
```

File: crates/pipeline-manager/src/api/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    error::ManagerError,
+};
+
+/// Returns the metrics of all running pipelines belonging to this tenant
+#[utoipa::path(
+    context_path = "/v0",
+    security(("JSON web token (JWT) or API key" = [])),
+    responses(
+        (status = OK
+        , description = "Returns the metrics of all running pipelines belonging to this tenant in Prometheus format."
```
Comment (snkas) [inline crates/pipeline-manager/src/api/metrics.rs:24]: "Returns the" and the dot at end can be left left out.

File: crates/pipeline-manager/src/api/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+use crate::{
+    api::ServerState,
+    db::{
+        storage::Storage as _,
+        types::{pipeline::PipelineStatus, tenant::TenantId},
+    },
+    error::ManagerError,
+};
+
+/// Returns the metrics of all running pipelines belonging to this tenant
```
Comment (snkas) [inline crates/pipeline-manager/src/api/metrics.rs:18]: "Retrieve the metrics ..... to this tenant."
(For consistency across all endpoint docs)

File: crates/pipeline-manager/src/api/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+pub(crate) async fn get_metrics(
+    state: WebData<ServerState>,
+    tenant_id: ReqData<TenantId>,
+) -> Result<HttpResponse, ManagerError> {
+    let pipelines = state.db.lock().await.list_pipelines(*tenant_id).await?;
+
+    const NEWLINE: u8 = b'\n';
+
+    let mut result = Vec::new();
+
```
Comment (snkas) [inline crates/pipeline-manager/src/api/metrics.rs:40]: Unnecessary empty line

File: crates/pipeline-manager/src/api/metrics.rs (hunk 3 above)
Comment (snkas) [reply crates/pipeline-manager/src/api/metrics.rs:18]: In addition, there should be some remark that these are fetched over time:
"The metrics are collected by making individual HTTP requests to the /metrics endpoint of each pipeline, of which only successful responses are included in the returned list."

File: crates/pipeline-manager/src/api/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    for pipeline in pipelines {
+        if pipeline.deployment_status == PipelineStatus::Running
+            || pipeline.deployment_status == PipelineStatus::Paused
+        {
+            if let Ok(res) = state
+                .runner
+                .forward_to_pipeline(*tenant_id, &pipeline.name, Method::GET, "metrics", "", None)
```
Comment (snkas) [inline crates/pipeline-manager/src/api/metrics.rs:47]: For now using the name is fine, after the runner rework I might change this to use pipeline identifier for it to be more robust to renaming.

**Review by snkas — APPROVED** [review]

**Comment by snkas** [comment]
Also could you add a test or provide a couple of bash instructions how to run the docker containers?

**Pull request merged**